# Анализ базы данных

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Компания решила быть на волне и купила крупный сервис для чтения книг по подписке. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

**Цель** - проанализировать базу данных, для помощи формирования ценностного предложения для нового продукта.

**Задания:**
* Посчитать, сколько книг вышло после 1 января 2000 года;
* Для каждой книги посчитать количество обзоров и среднюю оценку;
* Определить издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключим из анализа брошюры;
* Определить автора с самой высокой средней оценкой книг — учитывать только книги с 50 и более оценками;
* Посчитаь среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

**Ход исследования:**
1. Обзор данных
2. SQL-запросы
4. Вывод

## Шаг 1. Обзор данных

Импортируем необходимые библиотеки.

In [1]:
#!/usr/bin/python
# -*- coding: utf-8 -*-
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine
import datetime as dt

    # устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 



In [2]:
for table in ['books', 'authors', 'publishers', 'ratings', 'reviews']:
    query = ''' SELECT * FROM {}; 
    '''.format(table) 

    request = pd.io.sql.read_sql(query, con = engine) 
#     print(f'Таблица - "{table}"')
    display(request.head())
    print(request.info())
    print(f'Таблица - "{table}"')
    print('====================================================================================================')



book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB
None
Таблица - "books"


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB
None
Таблица - "authors"


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB
None
Таблица - "publishers"


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB
None
Таблица - "ratings"


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB
None
Таблица - "reviews"


В таблице `books`(содержит данные о книгах) 6 столбцов. Согласно документации к данным:
* `book_id` — идентификатор книги;
* `author_id` — идентификатор автора;
* `title` — название книги;
* `num_pages` — количество страниц;
* `publication_date` — дата публикации книги;
* `publisher_id` — идентификатор издателя.


В таблице `authors`(содержит данные об авторах) 2 столбца. Согласно документации к данным: 
* `author_id` — идентификатор автора;
* `author` — имя автора.

В таблице `publishers`(cодержит данные об издательствах) 2 столбца. Согласно документации к данным:
* `publisher_id` — идентификатор издательства;
* `publisher` — название издательства;

В таблице `ratings`(содержит данные о пользовательских оценках книг) 4 столбца. Согласно документации к данным:
* `rating_id` — идентификатор оценки;
* `book_id` — идентификатор книги;
* `username` — имя пользователя, оставившего оценку;
* `rating` — оценка книги.

В таблице `reviews`(cодержит данные о пользовательских обзорах на книги) 4 столбца. Согласно документации к данным:
* `review_id` — идентификатор обзора;
* `book_id` — идентификатор книги;
* `username` — имя пользователя, написавшего обзор;
* `text` — текст обзора.

Данные нуждаются в предобработке: изменение типа данных в столбце `publication_date` (таблица `books`)

Таблицы `books`, `authors`, `publishers`, `ratings`, `reviews` сохраним в одноименные переменные. 

## Шаг 3. SQL-запросы

1. Посчитайте, сколько книг вышло после 1 января 2000 года

In [3]:
query1 = ''' 
SELECT COUNT(*) 
FROM books
WHERE publication_date > '2000-1-1'; 
''' 

request1 = pd.io.sql.read_sql(query1, con = engine)


print(f'После 1 января 2000 года вышло {request1.iloc[0,0]} книг')

После 1 января 2000 года вышло 819 книг


2. Для каждой книги посчитайте количество обзоров и среднюю оценку

<div id='link2'></div>

In [4]:
query2 = ''' 

SELECT 
    books.title AS book_name,
    books.book_id AS id,
    SubRev.n_rev AS reviews,
    SubRat.avg_rat AS avg_rating
FROM books LEFT JOIN 
                (SELECT
                    book_id,
                    AVG(rating) AS avg_rat
                FROM
                    ratings
                GROUP BY 
                    book_id
                ) AS SubRat ON SubRat.book_id = books.book_id
        LEFT JOIN
                (SELECT 
                    book_id,
                    COUNT(review_id) AS n_rev
                FROM
                    reviews
                GROUP BY 
                    book_id
                ) AS SubRev ON SubRev.book_id = books.book_id
                
; ''' 

request2 = pd.io.sql.read_sql(query2, con = engine)
request2

book_name   id  reviews  \
0            The Body in the Library (Miss Marple  #3)  652      2.0   
1                                            Galápagos  273      2.0   
2                             A Tree Grows in Brooklyn   51      5.0   
3    Undaunted Courage: The Pioneering First Missio...  951      2.0   
4                                          The Prophet  839      4.0   
..                                                 ...  ...      ...   
995                                Alice in Wonderland   64      4.0   
996          A Woman of Substance (Emma Harte Saga #1)   55      2.0   
997                                          Christine  148      3.0   
998  The Magicians' Guild (Black Magician Trilogy  #1)  790      2.0   
999                           The Plot Against America  828      2.0   

     avg_rating  
0      4.500000  
1      4.500000  
2      4.250000  
3      4.000000  
4      4.285714  
..          ...  
995    4.230769  
996    5.000000  
997    3.428571  
998    3.500000  
999    3.000000  

[1000 rows x 4 columns]

3. Определите издательство, которое выпустило наибольшее число книг толще 50 страниц. Так мы исключим из анализа брошюры

In [5]:
query3 = ''' 
SELECT
    T.publisher_id,
    T.publisher,
    T.n_books
FROM
    (SELECT 
        books.publisher_id AS publisher_id,
        COUNT(books.book_id) AS n_books,
        publishers.publisher AS publisher 
    FROM 
        books 
        LEFT JOIN publishers ON books.publisher_id=publishers.publisher_id

    WHERE 
        num_pages>50
    GROUP BY
        books.publisher_id,
        publishers.publisher
    ORDER BY 
        n_books DESC) AS T
LIMIT 1
; ''' 

request3 = pd.io.sql.read_sql(query3, con = engine)
request3

publisher_id      publisher  n_books
0           212  Penguin Books       42

4. Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [6]:
query4 = ''' 
SELECT 
    aut_avg.author,
    AVG(aut_avg.avg_rating) AS avg_avg_rating
FROM
        (SELECT 
            authors.author AS author,
            AVG(ratings.rating) AS avg_rating
        FROM books
            LEFT JOIN authors ON books.author_id = authors.author_id
            LEFT JOIN ratings ON books.book_id = ratings.book_id
        GROUP BY 
             books.book_id,
             authors.author_id
        HAVING
            COUNT(ratings.rating_id) >=50 
        ) AS aut_avg 
GROUP BY    
    aut_avg.author
ORDER BY
    avg_avg_rating DESC
LIMIT 5
    
; ''' 

request4 = pd.io.sql.read_sql(query4, con = engine)
request4

author  avg_avg_rating
0         J.K. Rowling/Mary GrandPré        4.283844
1  Markus Zusak/Cao Xuân Việt Khương        4.264151
2                     J.R.R. Tolkien        4.258446
3                  Louisa May Alcott        4.192308
4                       Rick Riordan        4.080645

Самая высокая у J.K. Rowling/Mary GrandPré

5. Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [7]:
query5 = ''' 

SELECT
    AVG(Sub.n_reviews)
FROM
    (SELECT
        username,
        COUNT(review_id) AS n_reviews
    FROM
        reviews
    WHERE 
        username IN (
                SELECT 
                    username
                FROM 
                    ratings
                GROUP BY 
                    username
                HAVING 
                    COUNT(rating_id) > 50
                    )
    GROUP BY 
        username
    ) AS Sub
; ''' 

request5 = pd.io.sql.read_sql(query5, con = engine)
request5

avg
0  24.333333

## Вывод
1. Посчитайте, сколько книг вышло после 1 января 2000 года - `819 книг`
2. Для каждой книги посчитайте количество обзоров и среднюю оценку - <a href=#link2>ссылка</a>
3. Определите издательство, которое выпустило наибольшее число книг толще 50 страниц - `Penguin Books`
4. Определите автора с самой высокой средней оценкой книг — `J.K. Rowling/Mary GrandPré`
5. Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок - `24`

<div id='link1'></div> 